iroh-base/src/auth.rs

```diff
@@ -0,0 +1,93 @@
+//! Authentication related types and tooling.
+
+use std::future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::{future::Future, ops::Deref};
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+
+use crate::hash::Hash;
+
+/// The error code sent using quinn when aborting due to authentication errors.
+pub const REJECTED_CODE: u32 = 10;
+
+#[derive(Debug, Clone)]
+pub struct Authenticator(Arc<dyn DynAuthenticator>);
+
+impl Deref for Authenticator {
+    type Target = dyn DynAuthenticator;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+impl<A: DynAuthenticator> From<A> for Authenticator {
+    fn from(a: A) -> Self {
+        Authenticator(Arc::new(a))
+    }
+}
+
+type BoxFuture<T> = Pin<Box<dyn Future<Output = T> + Send + 'static>>;
+
+pub trait DynAuthenticator: Sync + Send + std::fmt::Debug + 'static {
```

Comment:
It would be good to have doc comments here to describe the flow. So for each request the request fn is called, which may or may not produce a token. And then response is called with the request and the token from above, and tells us whether to proceed. What is the purpose of the split compared to just having request -> ok/not ok?

Reply:
added some docs